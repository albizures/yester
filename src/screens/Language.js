import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, StyleSheet, Alert } from 'react-native'
import Container from '../components/Container'
import TopBar from '../components/TopBar'
import SettingsItem from '../components/SettingsItem'
import withUser, { shapeContextUser } from '../components/withUser'
import {
  updateUserAttribute,
  // cleanUserNotifications,
} from '../utils/session'
import { translate } from '../components/Translate'
import { screen } from '../utils/analytics'
import debugFactory from 'debug'

const debugError = debugFactory('yester:Language:error')
const debugInfo = debugFactory('yester:Language:info')

const isCheck = (locale, currentLocale) => {
  if (locale === currentLocale) {
    return SettingsItem.types.CHECK
  }
}

const LanguageItem = (props) => {
  const { currentLocale, locale, onPress } = props
  return (
    <SettingsItem
      {...props}
      type={isCheck(locale, currentLocale)}
      onPress={() => onPress(locale)}
    />
  )
}

LanguageItem.propTypes = {
  currentLocale: PropTypes.string.isRequired,
  locale: PropTypes.string.isRequired,
  onPress: PropTypes.func.isRequired,
}

class Language extends Component {
  static propTypes = {
    navigation: PropTypes.object.isRequired,
    contextUser: PropTypes.shape(shapeContextUser).isRequired,
  }

  constructor (props) {
    super(props)
    const { locale } = props.contextUser.user
    this.state = {
      locale,
      isLoading: false,
    }
    // Dev purposes, delete after test
    // cleanUserNotifications()
  }

  componentDidMount () {
    screen('Language', {})
  }

  onBack = () => {
    const { navigation } = this.props
    navigation.goBack()
  }

  onPress = async (locale) => {
    const {
      navigation,
      contextUser: { updateUser },
    } = this.props
    this.setState({ isLoading: true, locale })
    try {
      await updateUserAttribute('locale', locale)
      await updateUser()
      navigation.navigate('AppLoading', {
        lastScreen: 'Language',
      })
    } catch (error) {
      Alert.alert(translate('language.error'))
      debugError(error, error.message)
      this.setState({ isLoading: false })
    }
  }

  render () {
    const { isLoading, locale: currentLocale } = this.state
    const topBar = <TopBar title='language.title' onBack={this.onBack} />

    return (
      <Container topBar={topBar} isLoading={isLoading}>
        <View style={styles.container}>
          <LanguageItem
            title={translate('settings.language.item.en')}
            currentLocale={currentLocale}
            locale='en'
            onPress={this.onPress}
          />
          <LanguageItem
            title={translate('settings.language.item.es')}
            currentLocale={currentLocale}
            locale='es'
            onPress={this.onPress}
          />
        </View>
      </Container>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 20,
  },
  item: {
    paddingVertical: 20,
  },
})

export default withUser(Language)
