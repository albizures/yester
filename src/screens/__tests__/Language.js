import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { Button } from 'react-native'

import Language from '../Language'

const navigation = { navigate: jest.fn() }

describe('src/screens/Language.js', () => {
  it('should render', () => {
    const tree = renderer.create(<Language navigation={navigation} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('when button to home is pressed', () => {
    it('should navigate to Home screen', () => {
      const shallowLanguage = shallow(<Language navigation={navigation} />)
      const shallowTouchableHighlights = shallowLanguage.find(Button)
      const shallowButtonToHome = shallowTouchableHighlights.at(0)

      shallowButtonToHome.prop('onPress')()

      expect(navigation.navigate).toHaveBeenCalledTimes(1)
      expect(navigation.navigate).toHaveBeenCalledWith('Home')
    })
  })
})
