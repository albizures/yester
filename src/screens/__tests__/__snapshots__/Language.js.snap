// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/screens/Language.js should render 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#F6F7F9",
        "flex": 1,
      },
      undefined,
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1E1437",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 70,
        "paddingTop": 20,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      Language
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 75,
              "width": 340,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#313131",
                  "fontFamily": "Karla-Regular",
                  "textAlign": "left",
                },
                Object {
                  "fontSize": 18,
                },
                undefined,
              ]
            }
          >
            English
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "alignSelf": "center",
              "height": 35,
              "justifyContent": "center",
              "marginRight": 25,
              "width": 35,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/chevron/check.png",
              }
            }
            style={
              Object {
                "height": 9.5,
                "width": 13,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#D5D3DD",
              "borderColor": "#D5D3DD",
              "borderWidth": 0,
              "height": 1,
              "marginHorizontal": 22,
            },
            Object {
              "marginLeft": 0,
              "width": 323,
            },
          ]
        }
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 75,
              "width": 340,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#313131",
                  "fontFamily": "Karla-Regular",
                  "textAlign": "left",
                },
                Object {
                  "fontSize": 18,
                },
                undefined,
              ]
            }
          >
            Spanish
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#D5D3DD",
              "borderColor": "#D5D3DD",
              "borderWidth": 0,
              "height": 1,
              "marginHorizontal": 22,
            },
            Object {
              "marginLeft": 0,
              "width": 323,
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
